File: Examples/R/tests/testModels.R
test_that("Unconditional model exports are available", {
  classNames <- c(
    "FMTmodel",
    "FMTsemodel",
    "FMTsesmodel",
    "FMTsamodel"
  )

  for (className in classNames) {
    expect_true(
      exists(className, envir = asNamespace("FMT"), inherits = FALSE),
      info = paste("Missing R model export:", className)
    )
  }

  expect_true(
    exists(
      "FMTintmodelparameters",
      envir = asNamespace("FMT"),
      inherits = FALSE
    )
  )

  expect_true(
    exists(
      "FMTdblmodelparameters",
      envir = asNamespace("FMT"),
      inherits = FALSE
    )
  )

  expect_true(
    exists(
      "FMTboolmodelparameters",
      envir = asNamespace("FMT"),
      inherits = FALSE
    )
  )
})

test_that("OSI model exports follow compiled features", {
  skipIfFMTFeatureMissing("OSI")

  classNames <- c(
    "FMTlpsolver",
    "FMTsrmodel",
    "FMTnssmodel",
    "FMTlpmodel"
  )

  for (className in classNames) {
    expect_true(
      exists(className, envir = asNamespace("FMT"), inherits = FALSE),
      info = paste("Missing OSI-dependent R export:", className)
    )
  }

  skipIfSolverInterfaceMissing()
  solverInterface <- fmtSolverInterface()
  expect_false(is.null(solverInterface$CLP))
})

test_that("Mosek enum follows compiled features", {
  skipIfFMTFeatureMissing("MOSEK")
  skipIfSolverInterfaceMissing()

  solverInterface <- fmtSolverInterface()

  if (is.null(solverInterface$MOSEK)) {
    skip(
      paste(
        "FMT reports Mosek support, but MOSEK is not",
        "present in the exported solver interface"
      )
    )
  }

  expect_false(is.null(solverInterface$MOSEK))
})

File: Examples/Python/tests/testModels.py
import pytest
from FMT import Models, Parser
from conftest import requireFeature


def testUnconditionalModelExports():
    constructors = {
        "FMTmodel": Models.FMTmodel,
        "FMTsemodel": Models.FMTsemodel,
        "FMTsesmodel": Models.FMTsesmodel,
        "FMTsamodel": Models.FMTsamodel,
    }
    for name, constructor in constructors.items():
        assert constructor() is not None


def testModelParameterEnumsAreExposed():
    assert Models.FMTintmodelparameters.LENGTH is not None
    assert Models.FMTdblmodelparameters.TOLERANCE is not None
    assert Models.FMTboolmodelparameters.POSTSOLVE is not None


@pytest.mark.requiresOsi
def testOsiModelExports(modelPath):
    requireFeature("OSI")
    assert hasattr(Models, "FMTsolverinterface")
    assert hasattr(Models, "FMTlpmodel")
    parser = Parser.FMTmodelparser()
    parser.setquietlogger()
    models = parser.readproject(str(modelPath), ["LP"])
    model = Models.FMTlpmodel(models[0], Models.FMTsolverinterface.CLP)
    assert model is not None
    assert len(model.getoutputs()) > 0


@pytest.mark.requiresMosek
def testMosekEnumMatchesCompiledFeature():
    requireFeature("MOSEK")
    assert hasattr(Models.FMTsolverinterface, "MOSEK")
